### DISCUSSION TITLE
How to read delta lake table via an s3 bucket

### DISCUSSION THREAD

**Discussion by finlaydotb** [delta-io/delta-rs#1165] [Q&A] [answered]
As seen from the documentation [here](https://docs.rs/deltalake/0.7.0/deltalake/) it is possible to use the `deltalake::open_table` method to open the delta lake table located on the file system. This I have been able to do.

 But I have not been able to figure out how to open delta lake table on s3. 

The documentation for `deltalake::open_table` which can be found [here](https://docs.rs/deltalake/0.7.0/deltalake/delta/fn.open_table.html) says

> Creates and loads a DeltaTable from the given path with current metadata. Infers the storage back-end to use from the scheme in the given table path.

I tried to pass in a url to s3 bucket since the documentation says the storage back-end will be inferred. But doing this `open_table("s3://127.0.0.1:9000/deltalake/delta-table")` leads to this panic

> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: LoadCheckpoint { source: Storage { source: Generic { store: "S3", source: Error { retries: 0, message: "request error", source: Some(reqwest::Error { kind: Request, url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(169.254.169.254)), port: None, path: "/latest/api/token", query: None, fragment: None }, source: hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 60, kind: TimedOut, message: "Operation timed out" })) }) } } } }', src/main.rs:71:10
stack backtrace:
   0: rust_begin_unwind

I had a feeling this should be trivial but unfortunately it has not been :(

So anyone knows how to get open_table to open a table located on an s3 bucket?

**Comment by roeap** [comment, marked as answer]
Hi @finlaydotb,

Working against an emulated service requires a bit more configuration, and we should update the docs to reflect that. Specifically, you can either set some configuration in the environment, or pass it explitily by using `open_table_with_storage_options`.

Looking at your example code, it would likely look something like this.

```rs
storage_options = HashMap::from([
    ("allow_http", "true"),
    ("endpoint_url", "http://127.0.0.1:9000"),
    ("access_key_id", "some id"),
    ("access_key_secret", "some secret")
]);
```

The actual error you are seeing is a different one though. Essentially the underlying object store tries to use the the metadata endpoint to fetch an access token, which is the default, if no credential is provided. If the local service allows anonymous access, an empty key id / secret should still be provided, to not query the metadata endpoint.

If you want to set the config in the environment, you can use the same kieys as above, just capitalized and prefixed with `AWS_`. A list of all supported config for S3 stores can be found [here](https://docs.rs/object_store/0.5.4/object_store/aws/enum.AmazonS3ConfigKey.html#variants).

Last but not least, the URL should be in the form of `s3://<bucket>/<path>`, which in your case would be `s3://deltalake/detla-table`.

**Reply by finlaydotb** [reply]
Hi @roeap 

Thanks for the answer. I updated the code with the essential part reproduced below:

```
    let storage_options = HashMap::from([
        ("allow_http".to_string(), "true".to_string()),
        ("endpoint_url".to_string(), "http://127.0.0.1:9000".to_string()),
        ("access_key_id".to_string(), "minioadmin".to_string()),
        ("access_key_secret".to_string(), "minioadmin".to_string())
    ]);

    let table = deltalake::delta::open_table_with_storage_options("s3://deltalake/delta-table", storage_options)
        .await
        .unwrap();
```

but this panicked with the following error:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
     Running `target/debug/analysis`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: MissingFeature { feature: "s3", url: "s3://deltalake/delta-table" }', src/main.rs:77:10
stack backtrace:
   0: rust_begin_unwind

```

So I thought to enable the s3 features. My updated cargo file looks like this:

```
[dependencies]
datafusion = "17.0.0"
deltalake = {version="0.7.0", features = ["datafusion", "s3"]}
tokio = {version="1.25.0", features = ["macros", "rt", "parking_lot"]}
```

Note the datafusion versio is set to `17.0.0` instead of latest which is `18` due to deltalake have dependency for datafusion set at `^17`

Unfortunately this fails to compile with error

```
error[E0308]: mismatched types
    --> /path/to/user/.cargo/registry/src/github.com-1ecc6299db9ec823/deltalake-0.7.0/src/storage/s3.rs:527:17
     |
526  |             let lock_client = dynamodb_lock::DynamoDbLockClient::new(
     |                               -------------------------------------- arguments to this function are incorrect
527  |                 dynamodb_client,
     |                 ^^^^^^^^^^^^^^^ expected struct `rusoto_dynamodb::generated::DynamoDbClient`, found struct `DynamoDbClient`
```

Looks like issue with dependencies lining up correctly?

**Reply by roeap** [reply]
hmm ... I would have to investigate that a bit more. Could you try patching the `dynamodb_lock` dependency to point at the repo and see if that compiles?

Also, since you are local, you may also set `AWS_S3_ALLOW_UNSAFE_RENAME` to "true" in your environment. This is really only needed in multi-.writer scenarios, and allows you to avoid setting up a local dynamodb.

**Reply by finlaydotb** [reply]
> Could you try patching the `dynamodb_lock` dependency to point at the repo and see if that compiles?


So first time doing this so I had to look up how to...so I checked out the repo https://github.com/delta-io/delta-rs, then modified my Cargo.toml to 

```
[patch.crates-io]
dynamodb_lock = {path = "./delta-rs/dynamodb_lock"}
```

This compiled fine. (curious is there a shorter way to patch dependencies?)

I tried your second suggestion by running `AWS_S3_ALLOW_UNSAFE_RENAME="true" cargo build` It seems not to remove the need for `dynamodb` as the compilation still failed with it.

**Reply by roeap** [reply]
> curious is there a shorter way to patch dependencies?

You can directly reference a git repo

```toml
[patch.crates-io]
dynamodb_lock = {git = "https://github.com/delta-io/delta-rs", rev="..."}
```

for the second part I was a little ambiguous :). Compilation is still needed and the parameter will not help during build. Just when using the S3 there is also a Dynamo DB required, since S3 does not support atomic renames. If there is a single writer though, we can live without atomic renames, and the option `AWS_S3_ALLOW_UNSAFE_RENAME` allows you to avoid instantiating the dynamo client at runtime...

**Reply by finlaydotb** [reply]
Thanks! I was not sure about using the git route for the patch, as the url points to the root of the project not specifically to `dynamodb_lock`, but with the `path` route, the path given actually points to `dynamodb_lock`'s directory and where it's Cargo.toml file is located.

So I guess having this path is a workable workaround? Because it seems the `dynamodb_lock` on crates.io is stuck at version `0.4.2` - is the plan to stop publishing `dynamodb_lock` to crates.io?

**Reply by roeap** [reply]
We definitely want to keep publishing `dynamodb_lock`! There was a larger recent refactor concerning building the repository, where we needed to pin the rusoto dependencies, which is likely the root cause. @houqp - can we do another release for `dynamodb_lock`?

**Comment by vegarsti** [comment]
In case anyone needs this: We need to also call `deltalake::aws::register_handlers(None)`
